File: 백준/Silver/2644. 촌수계산/촌수계산.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace CodingTest
{
    internal class Program
    {
        static bool[] visited;
        static List<int>[] list;
        static int answer;
        static int b;
        static void DFS(int node, int count)
        {
            visited[node] = true;
            for (int i = 0; i < list[node].Count; i++)
            {
                int x = list[node][i];
                if (!visited[x])
                {
                    if (x == b)
                    {
                        answer = count + 1;
                        return;
                    }
                    DFS(x, count + 1);
                }
            }
        }
        static void Main(string[] args)
        {
            int n = int.Parse(Console.ReadLine());
            visited = new bool[n + 1];
            list = new List<int>[n + 1];
            for (int i = 0; i <= n; i++)
            {
                list[i] = new List<int>();
            }

            string[] s = Console.ReadLine().Split();
            int a = int.Parse(s[0]);
            b = int.Parse(s[1]);

            int m = int.Parse(Console.ReadLine());
            for (int i = 0; i < m; i++)
            {
                string[] input = Console.ReadLine().Split();
                list[int.Parse(input[0])].Add(int.Parse(input[1]));
                list[int.Parse(input[1])].Add(int.Parse(input[0]));
            }

            answer = -1;
            DFS(a, 0);
            Console.WriteLine(answer);
        }
    }
}

File: 백준/Silver/2644. 촌수계산/촌수계산.cc
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int n, m;
int a, b;
vector<vector<int>> graph;
vector<bool> visited;
int answer = -1;

void dfs(int x, int count) {
    visited[x] = true;

    for (int i = 0; i < graph[x].size(); i++)
    {
        int y = graph[x][i];
        if (!visited[y])
        {
            if (y == b)
            {
                answer = count + 1;
                return;
            }
            dfs(y, count + 1);
        }
    }
}

int main(void) {
    cin >> n;
    cin >> a >> b;
    cin >> m;
    graph.resize(n + 1);
    visited.resize(n + 1);
    for (int i = 0; i < m; i++)
    {
        int p, q;
        cin >> p >> q;
        graph[p].push_back(q);
        graph[q].push_back(p);
    }

    dfs(a, 0);
    cout << answer << " ";

    return 0;
}